crates/sui-network/src/state_sync/mod.rs

```diff
@@ -1046,6 +1068,75 @@ where
     Ok(())
 }
 
+async fn sync_checkpoint_contents_from_archive<S>(
+    network: anemo::Network,
+    archive_readers: ArchiveReaderBalancer,
+    store: S,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    metrics: Metrics,
+) where
+    S: WriteStore + Clone + Send + Sync + 'static,
+    <S as ReadStore>::Error: std::error::Error,
+{
+    loop {
+        let peers: Vec<_> = peer_heights
+            .read()
+            .unwrap()
+            .peers_on_same_chain()
+            // Filter out any peers who we aren't connected with.
+            .filter_map(|(peer_id, info)| network.peer(*peer_id).map(|peer| (peer, *info)))
+            .collect();
+        let lowest_checkpoint_on_peers = peers
+            .iter()
+            .map(|(_p, state_sync_info)| state_sync_info.lowest)
+            .min();
+        let highest_synced = store
+            .get_highest_synced_checkpoint()
+            .expect("store operation should not fail")
+            .sequence_number;
+        let sync_from_archive = if let Some(lowest_checkpoint_on_peers) = lowest_checkpoint_on_peers
+        {
+            highest_synced < lowest_checkpoint_on_peers
+        } else {
+            false
+        };
+        if sync_from_archive {
+            let start = highest_synced
+                .checked_add(1)
+                .expect("Checkpoint seq num overflow");
+            if let Some(archive_reader) = archive_readers
+                .pick_one_random(start..lowest_checkpoint_on_peers.unwrap())
```

Comment:
This is to load balance across archive backends actually (each archive_reader proxies a bucket). A node (mysten run or community) might have access to multiple archive buckets i.e. S3, GCS, etc which are being maintained by different parties (A community node might have access to its own archives as well as mysten operated ones). In such a scenario, we just pick one of the random buckets to fulfill the incoming read request

Reply:
i understand, but i think load balancing across entirely different services like this is going to lead to very unpredictable behavior if any of them are having problems or outages. It also seems counter to how people would actually want to use the service. For instance if they have buckets in different regions, they might want to have nodes in each region have the nearby archive as a primary, and fall back to the secondary as a backup. Its hard for me to imagine that they'd want to randomly choose a backend.

But, this is not a big deal and we can just leave it as is until we need to change it.

Reply:
This is truly the most trivial load balancer we can have. I didn't put much thought into it because we just want something for now and it's unclear what the performance profile of different archive services might look like under load. I imagine when we set up an archive bucket on different cloud services we can choose max throughput, max concurrent downloads, etc all of which is a variable to consider in designing a good load balancer (which could be primary-secondary, weighted, or something smarter even).

For now, if any of the backends are having problems or outages, retrying should find them a working backend (not optimal but works). And for predictable performance we could probably advise operators to have a small list of high quality backends in their config.